File: src/Frontend/FluentCMS.Web.UI.Components/Components/Form/FormTreeSelector/FormTreeSelector.razor.cs
using System.Diagnostics.CodeAnalysis;

namespace FluentCMS.Web.UI.Components;

public partial class FormTreeSelector : IAsyncDisposable
{
    [Inject]
    public IJSRuntime? JS { get; set; }

    [Parameter]
    public int Cols { get; set; } = 12;

    [Parameter]
    public EventCallback<string?> OnChange { get; set; }

    [Parameter]
    public List<TreeSelectorItemType> Items { get; set; } = [];

    public ElementReference? Element { get; set; }

    private IJSObjectReference? Module { get; set; }

    private DotNetObjectReference<FormTreeSelector>? DotNetRef { get; set; }

    private string? _value;

    [JSInvokable]
    public async Task UpdateValue(string value)
    {
        _value = value;
        await ValueChanged.InvokeAsync(value);
        await OnChange.InvokeAsync(value);
    }

    protected override async Task OnParametersSetAsync()
    {
        if (Value == _value) return;

        _value = Value;

        if (Module != null)
            await Module.InvokeVoidAsync("update", DotNetRef, Element, new { Value });
    }

    public async ValueTask DisposeAsync()
    {
        if (Module is not null)
        {
            await Module.InvokeVoidAsync("dispose", DotNetRef, Element);
            await Module.DisposeAsync();
        }
        DotNetRef?.Dispose();
    }

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (!firstRender) return;

        if (JS is null)
        {
            throw new InvalidOperationException("JS runtime has not been initialized.");
        }

        DotNetRef = DotNetObjectReference.Create(this);
        Module = await JS.InvokeAsync<IJSObjectReference>("import", "/_content/FluentCMS.Web.UI.Components/Components/Form/FormTreeSelector/FormTreeSelector.razor.js");
        await Module.InvokeVoidAsync("initialize", DotNetRef, Element, new { Items, Value });
    }

    protected override bool TryParseValueFromString(string? value, out string? result, [NotNullWhen(false)] out string? validationErrorMessage)
    {
        result = value;
        validationErrorMessage = null;
        return true;
    }
}

File: src/Frontend/FluentCMS.Web.UI.Components/Components/Form/FormTreeSelector/FormTreeSelector.razor.js
import {TreeSelector} from '/_content/FluentCMS.Web.UI.Components/js/tree-selector.js'
import { debounce } from '/_content/FluentCMS.Web.UI.Components/js/helpers.js'

const treeSelectors = new Map();

export function update(dotnet, element, config) {
    const instance = treeSelectors.get(element)

    if ('value' in config) {
        instance.setValue(config.value);
    }
}

export function initialize(dotnet, element, config) {
    dispose(dotnet, element);

    let instance = TreeSelector(element, {
        options: config.items,
        cssPrefix: 'f-tree-selector',
        value: config.value,
        onChange(value) {
            dotnet.invokeMethodAsync('UpdateValue', value)
        },
    })

    treeSelectors.set(element, instance);
}

export function dispose(dotnet, element) {
    const treeSelector = treeSelectors.get(element);

    treeSelector?.dispose();

    treeSelectors.delete(element);
}
